### ISSUE TITLE
indicator for fast FMA

### ISSUE THREAD

**Issue by simonbyrne** [JuliaLang/julia#9855]
Now we have a `fma` function, it would be useful to have some method for determining whether or not this is more efficient than the naive `x*y+z`, particularly for use in algorithms that use double-double style arithmetic (C provides the `FP_FAST_FMA` macro for this purpose).

From https://github.com/JuliaLang/julia/pull/8112#issuecomment-66527501, it seems that the best option is to expose `TargetLowering::isFMAFasterThanFMulAndFAdd` (presumably in `base/sysinfo.jl`?)

**Comment by ivarne** [comment]
Isn't this the problem `muladd` from https://github.com/JuliaLang/julia/pull/9840 is intended to solve?

**Comment by simonbyrne** [comment]
Not quite: they are related but distinct.

* The purpose `muladd` is to compute `x*y+z` in the fastest way possible (i.e. in a Horner evaluation scheme for a polynomial).
* The purpose of this is to determine whether `fma` is fast (i.e. in hardware), or slow (i.e. in software) in which case you might want to use a different approach altogether. An example is [here](https://github.com/simonbyrne/libm.jl/blob/d6678d255bf5a019d0d7493e34bad7a6a731ac86/src/log_tang.jl#L105-L113): if you have a hardware `fma`, it can reduce to 5 operations, otherwise you need 12 (using a software `fma` would most likely be even slower).

**Comment by simonbyrne** [comment]
As a rough point of reference, a software `fma` is about 10x slower than a (non-fused) multiply-add on my computer.

**Comment by nalimilan** [comment]
Couldn't a function be designed so that it would automatically use the fastest solution? Maybe I'm being naive though.

**Comment by simonbyrne** [comment]
@nalimilan I'm not sure what you mean: do you mean something like specifying two code paths and letting the compiler pick the one it likes the most?

**Comment by nalimilan** [comment]
@simonbyrne No, I wonder whether `u+muladd(fma(-u,f,2(f-u)),g,q)` couldn't be automatically translated to the detailed steps you show [here](https://github.com/simonbyrne/libm.jl/blob/d6678d255bf5a019d0d7493e34bad7a6a731ac86/src/log_tang.jl#L107-L113) if `fma` is known to be slow. But I guess it's more complex than that.

**Comment by simonbyrne** [comment]
Ah I see: you could do that, but it would probably be around twice as slow as the alternative: a decent software `fma` (e.g. [openlibm](https://github.com/JuliaLang/openlibm/blob/master/src/s_fma.c)) does a few extra operations to avoid things like overflow and double rounding, which are not needed in this particular case.

**Comment by eschnett** [comment]
Another point: `fma` is guaranteed to avoid rounding the intermediate result, which improves accuracy, and this is what allows alternative algorithms to be used. `muladd`, on the other hand, may be fast, but may still round the intermediate result. For example, ARM has both `vmla` and `fma` instructions -- the former rounds the intermediate result, the latter doesn't.

**Comment by JeffreySarnoff** [comment]
When using double-double algorithms for extended precision math or to get Float64 results from Float32 friendly GPUs, `fma` is essential (it is possible to do without -- but everything takes much, much longer and gets more complicated).  It is that the fma only-rounds-once that matters for this.

After trying it both ways, I am using fma everywhere possible -- whether or not the fma is slow (software emulated); the alternative is backwards facing and confounding for careful numerics.

**Comment by musm** [comment]
how can `TargetLowering::isFMAFasterThanFMulAndFAdd`  be exposed, its a cpp fun?

**Comment by eschnett** [comment]
The canonical way would be to either introduce an intrinsic function that returns a `Bool`, or to provide a `C` wrapper that can be called via `ccall`. Both require changes to Julia's core.

**Comment by simonbyrne** [comment]
There is currently an attempt to do this in the code via comparing `muladd` to `fma`, however it doesn't seem to work:
https://github.com/JuliaLang/julia/blob/4a046009a3362ab5e17d369641dbbc9657eb680c/base/special/log.jl#L144
as on my machine (which does have FMA) I get:
```
julia> Base.Math.FMA_NATIVE
false
```

**Comment by JeffreySarnoff** [comment]
The sign is wrong, try this:
```
const FMA_NATIVE = muladd(nextfloat(1.0),nextfloat(1.0),-nextfloat(1.0,2)) == 4.930380657631324e-32
```

**Comment by simonbyrne** [comment]
🤦‍♂

**Comment by simonbyrne** [comment]
(fixed in #32318)

**Comment by simonbyrne** [comment]
I'll close this for now, move all discussion to #33011.
